### ISSUE TITLE
Recover geth and story clients when geth client is upgraded after the upgrade height

### ISSUE THREAD

**Issue by limengformal** [piplabs/story#144]
## Description and context
In the event of a geth hard fork upgrade, if a node is not upgraded after the upgrade block height, the story client may panic since it may not reach consensus with the rest of the nodes that already have their geth clients upgraded.

Only upgrading the geth client at this point doesn't help since the story client already verified/proposed a block that is in conflict. Node at this point can only remove data folder and sync from genesis block which takes a long time.

## Suggested solution
One suggested solution is to roll back the block with incorrect proposal so the node can sync the correct block.

## Definition of done
Node can roll back incorrect block and restart with correct block in the event of a late geth upgrade.

**Comment by limengformal** [comment]
How to reproduce:

- versions
  - before upgrade: story client version / geth version
    - story: v0.10.0
    - geth: v0.9.2
  - after upgrade, normal node: story client version / geth version
    - story: v0.10.0
    - geth: v0.9.3
  - after upgrade, not-upgrading node: story client version / geth version
    - story: v0.10.0
    - geth: v0.9.2
- upgrade procedure
  - Before the upgrade height (you may specify height by using --override.nostoi ${BLOCK_NUMBER})
  - Stop story
  - Stop old geth
  - Start new geth with upgrade height by using --override.nostoi ${BLOCK_NUMBER}
  - Start story
- Trigger the issue
  - Leave some nodes not upgraded to geth v0.9.3 after specified ${BLOCK_NUMBER}
  - Send a tx to call EIP-7212 precompile

Install Cast (if not installed yet):
```
curl -L https://foundry.paradigm.xyz/ | bash
source /home/ec2-user/.zshenv
foundryup
```

```
cast call 0x0000000000000000000000000000000000000100  "0x4cee90eb86eaa050036147a12d49004b6b9c72bd725d39d4785011fe190f0b4da73bd4903f0ce3b639bbbf6e8e80d16931ff4bcf5993d58468e8fb19086e8cac36dbcd03009df8c59286b162af3bd7fcc0450c9aa81be5d10d312af6c66b1d604aebd3099c618202fcfe16ae7770b0c49ab5eadf74b754204a3bb6060e44eff37618b065f9832de4ca6ca971a7a1adc826d0f7c00181a5fb2ddf79ae00b4e10e" --rpc-url <node url to 8545 port, e.g. http://localhost:8545/>

```

**Comment by zsystm** [comment]
@limengformal
The rollback command has been successfully integrated (#157) and is functioning as expected! Below is a brief summary of the steps taken, along with a video demonstration showing it in action.

https://github.com/user-attachments/assets/a038a174-29a2-4db0-83b1-90c8c8478663

### Video Explanation

**Background Info**
1. The terminal sessions on the left represent Geth sessions, listed from top to bottom as `N0, N1, N2,` and `N3`. Note that `N1` has not been upgraded and is currently using version [story-geth-v0.9.2](https://github.com/piplabs/story-geth/releases/tag/v0.9.2).
2. The terminal sessions on the right represent Story sessions, also listed from top to bottom as `N0, N1, N2,` and `N3`.

**Stuck Resolution Flow in the Video**
1. `N1`, which has not been upgraded, has a different app hash from the other nodes, causing it to get stuck and preventing further block progress. Meanwhile, the blocks on `N0, N2,` and `N3` are continuing to increase without issues.
2. To resolve the stuck issue on `N1`, I attempted a rollback on the story client for `N1` using the command: `./story rollback --home val1 --hard`.
3. The message `Rolled back state to ...` confirms that a rollback of one block was successfully performed.
4. After the rollback, I upgraded `N1`’s Geth from version v0.9.2 to [v0.9.3](https://github.com/piplabs/story-geth/releases/tag/v0.9.3) and restarted it.
5. I also restarted the story client for `N1`.
7. As seen in the second terminal sessions, the stuck issue is resolved, and `N1` successfully syncs to the latest block.

**Comment by zsystm** [comment]
> @limengformal I have successfully reproduced the stuck scenario.
> 
> * Note: Nodes N0 through N4 have equal voting power.
>   ![image](https://private-user-images.githubusercontent.com/124245155/372129527-aa6fcb8d-34ed-43fc-b7f5-84fdd7aaa2cf.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3Mjc3NjQyMTUsIm5iZiI6MTcyNzc2MzkxNSwicGF0aCI6Ii8xMjQyNDUxNTUvMzcyMTI5NTI3LWFhNmZjYjhkLTM0ZWQtNDNmYy1iN2Y1LTg0ZmRkN2FhYTJjZi5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjQxMDAxJTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI0MTAwMVQwNjI1MTVaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT05MGM2ZDdmNjEzNWZmYTdmN2Q5M2JkMTQxZmE2MzJhYWQwYTc1YWUyMjYyMWQ2NjliNGFmMzc0ZWRkYmNkZDIwJlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCJ9.qh4bmkN3wZKfEBg20eMCzz1AjnPRIQQ_BZgeRF4YuJE)
> 
> Currently, I'm working on integrating the CometBFT rollback command into the Story codebase to test if this rollback feature can resolve the issue.
> 
> * Since the Story codebase has some differences from a standard Cosmos SDK chain, I'm unable to directly apply the existing [rollback command](https://github.com/cosmos/cosmos-sdk/blob/6a0358607411cd83d22c6f1deea69f399f803007/server/rollback.go#L13-L14).
> * Integration is working on in [zsystm/rollback](https://github.com/zsystm/story/tree/zsystm/rollback) branch
> 
> geth logs at stuck
> story logs at stuck
> **Note:** Geth did not insert the invalid block, so it remains at the last valid block. This means that once the app hash issue in the story is resolved and re-executed, the chain would function correctly again.

What happens if a non-upgraded node proposes a block? No need to worry—other nodes will reject the proposal, and Geth will not insert the incorrect block into the database. 

In this case, the non-upgraded node should stop both the Story client and Geth, upgrade Geth, and then restart everything. 
If the node was in a stuck, it can perform a rollback first. If there is no stuck state, simply upgrading and restarting should allow the node to sync blocks without any issues.

**Comment by limengformal** [comment]
@zsystm Thank you for the work. This solution rollback a block on the CL side. Do we also need to rollback blocks on the EL side as well?

**Comment by zsystm** [comment]
@limengformal 
No, we don’t need to roll back blocks on the EL side in this case. Some nodes didn’t upgrade Geth in time meaning the hard fork is already active. Any non-upgraded node will stop processing blocks containing state-breaking transactions (i.e., transactions that result in different states due to the hard fork).

**Case 1: Non-upgraded node is the proposer**
- The EVM payload created by the non-upgraded node will be rejected by the other nodes. This means the payload will never be finalized on the CL side, and the non-upgraded node will not insert the block since it hasn’t been finalized by the CL.
**Case 2: Non-upgraded node is not the proposer**
- The non-upgraded node will reject the EVM payload created by upgraded nodes because it will differ from its own state.

In other scenarios, there may be a need for rollback. For example, if there were an attempt by a hacker to steal a large sum of funds on the EL side and the community, along with the majority of validators, deemed the action unjust, a rollback might be necessary. However, in the current situation where we are dealing with some nodes (less than 1/3 of total voting power) missing their upgrades, there is no need for a rollback on the EL side.

**Comment by limengformal** [comment]
@zsystm submitted [PR](https://github.com/piplabs/story/pull/157) to resolve the issue. Rollback feature is available now.
